isisd/isisd.c

```diff
@@ -3249,13 +3249,73 @@ void isis_area_overload_on_startup_set(struct isis_area *area,
 }
 
 void isis_area_advertise_high_metrics_set(struct isis_area *area,
-				       bool advertise_high_metrics)
+					  bool advertise_high_metrics, int restart_metric)
 {
 	if (area->advertise_high_metrics != advertise_high_metrics) {
-		area->advertise_high_metrics = advertise_high_metrics;
+		if (advertise_high_metrics) {
+			struct listnode *node;
+			struct isis_circuit *circuit;
+			int max_metric;
+
+			if (!restart_metric) {
+				if (area->oldmetric && area->newmetric) {
+					max_metric = ISIS_NARROW_METRIC_INFINITY;
+				} else if (area->newmetric) {
+					max_metric = MAX_WIDE_LINK_METRIC;
+				} else {
+					max_metric = MAX_NARROW_LINK_METRIC;
+				}	
+			} else {
+				max_metric = restart_metric;
+			}
 
-		// TODO: Implement functionality
+			for (ALL_LIST_ELEMENTS_RO(area->circuit_list, node,
+						  circuit)) {
+				isis_circuit_metric_set(circuit, IS_LEVEL_1,
+							max_metric);
+				isis_circuit_metric_set(circuit, IS_LEVEL_2,
+							max_metric);
+			}
+
+			area->advertise_high_metrics = advertise_high_metrics;
+			area->advertise_high_metric_value = max_metric;
+		} else {
+			struct listnode *node;
+			struct isis_circuit *circuit;
+
+			area->advertise_high_metrics = advertise_high_metrics;
+			for (ALL_LIST_ELEMENTS_RO(area->circuit_list, node,
+						  circuit)) {
+				char xpath[XPATH_MAXLEN];
+				struct lyd_node *dnode;
+				int configured_metric_l1;
+				int configured_metric_l2;
+
+				/* Get configured metric */
+				snprintf(
+					xpath, XPATH_MAXLEN,
+					"/frr-interface:lib/interface[name='%s']",
+					circuit->interface->name);
+				dnode = yang_dnode_get(running_config->dnode,
+						       xpath);
+
+				configured_metric_l1 = yang_dnode_get_uint32(
+					dnode,
+					"./frr-isisd:isis/metric/level-1");
+				configured_metric_l2 = yang_dnode_get_uint32(
+					dnode,
+					"./frr-isisd:isis/metric/level-2");
+
+				isis_circuit_metric_set(circuit, IS_LEVEL_1,
+							configured_metric_l1);
+				isis_circuit_metric_set(circuit, IS_LEVEL_2,
+							configured_metric_l2);
+
+				area->advertise_high_metric_value = 0;	
+			}
+		}
 	}
+	isis_restart_write_advertise_high_metric(area, area->advertise_high_metric_value);
```

Comment:
why do we need to do that if nothing has changed ?
Why don't we only store `advertise_high_metric` boolean in the non volatile file ? The value will be deducted from `area->new_metric` and `area->old_metric`

Reply:
Updated, removed non volatile file functionality